### PR TITLE
Basic Authentication with Strimzi Credentials

### PR DESCRIPTION
**Pull request by snophey**
This PR extends the registry with a `StrimziIdentityProvider` which, when enabled, will take the basic auth credentials from an incoming request and validate them against the relevant Strimzi-generated credentials inside the configured Kubernetes namespace. For more context, see #4724

### CODE REVIEW

**Comment by carlesarnal** [comment]
@snophey First of all thanks for the PR. Sorry, I'm back from PTO, you've got a couple conflicts, do you mind resolving them? Thanks

File: app/pom.xml [hunk 1]
```diff
@@ -189,6 +193,16 @@
             <artifactId>lombok</artifactId>
             <scope>compile</scope>
         </dependency>
+        <dependency>
+            <groupId>org.bouncycastle</groupId>
+            <artifactId>bcprov-jdk18on</artifactId>
+            <version>1.78.1</version>
```
Comment (carlesarnal) [inline app/pom.xml:199]: Versions must be defined in the main pom, not in the application module. We do it this way to keep consistency and avoid conflicts across the application

**Review by carlesarnal — CHANGES_REQUESTED** [review]
Apart from a small comment regarding versions, lgtm.

**Comment by andreaTP** [comment]
Hi @snophey , sorry for being late in tis review, I have a few questions about your proposed implementation:
- I miss seeing the extraction of roles, do you have any plan to support it? or only basic authorization?
- To perform the Secret reading we need additional rbac policies, how does it play with the operator?

I do see a couple of alternative implementations that would provide a similar functionality without having to directly contact the k8s API:
- mount the secret and use the built-in basic auth mechanism (discouraged in production)
- handle this feature in the Operator

I would be a bit conscious of directly contacting the K8s API from Registry as it brings in another bag of concerns (being able to contact the k8s API server, rbac etc. etc.).

cc. @EricWittmann @jsenko

**Comment by snophey** [comment]
Thank you for the reviews, @carlesarnal and @andreaTP. The merge conflicts should be resolved for the time being. I have also moved out the dependency versions to the parent pom.

> I miss seeing the extraction of roles, do you have any plan to support it? or only basic authorization?

The focus of this PR is to add better integration for Strimzi users with SCRAM-SHA-512 credentials and ACL-based permissions. To my knowledge, this setup does not have anything that could mapped to a role.

Users who want to use RBAC with their Kafka clients probably already have some sort of OAuth-setup with Keycloak running. At that point, they are probably best off also configuring their registry with Keycloak authorization.

> To perform the Secret reading we need additional rbac policies, how does it play with the operator?

I have not tested this version with the operator yet. However, generally we rely on the configuration settings from the `kubernetes-client` Quarkus extension (which builds on the fabric8 client library). This means that we can supply the credentials for accessing the K8s API in a [multitude of ways](https://quarkus.io/guides/kubernetes-client). For example we could inject a token as an environment variable or (more elegantly) execute the container with a service account that has the role bindings necessary for reading secrets. I am not very familiar with the operator, but if I remember correctly, the operator allows configuring the application's service account, so the latter approach should work.

> I do see a couple of alternative implementations that would provide a similar functionality without having to directly contact the k8s API:
> * mount the secret and use the built-in basic auth mechanism (discouraged in production)
> * handle this feature in the Operator

I suppose this would eliminate the need to talk to the k8s API, but it wouldn't it also require us to rollout new schema replicas whenever a user is added (in order to inject the updated configuration into the pod)? This could be a hard sell for some of our technical stakeholders.

**Comment by andreaTP** [comment]
Thanks for getting back @snophey !

> This means that we can supply the credentials for accessing the K8s API in a [multitude of ways](https://quarkus.io/guides/kubernetes-client).

(Fabric8 K8s client maintainer here) That's very correct, and this is a complexity I would love to keep away from Registry.

> I suppose this would eliminate the need to talk to the k8s API, but it wouldn't it also require us to rollout new schema replicas whenever a user is added (in order to inject the updated configuration into the pod)?
 
 A few observations here:
 - operators usually watch "managed" Secrets and they reconcile accordingly
 - making calls to K8s API for each and every login request might have security considerations
 - relying on a "local file"(as seen from Registry) makes the mechanism robust and doesn't break levels of abstractions

Thoughts?

**Comment by snophey** [comment]
> * operators usually watch "managed" Secrets and they reconcile accordingly

Could you elaborate on this a bit? My understanding is that if I were to go and create a `KafkaUser`, the Strimzi entity operator would create a "managed" secret with that user's credentials. In your proposal, the registry operator would detect this managed secret. What would a potential reconciliation look like? I can only think of the operator generating a new "passwords file" for the registry to read from (by collecting all Strimzi-managed secrets). But the operator cannot mount this updated secret unless it restarts the existing registry pods.

One of the things we appreciate about the registry is that it preserves schemas in-memory even if the Kafka backend becomes unavailable (thus being able to serve read-requests for the duration of the incident). Therefore, I think it would be great if a user change/addition/deletion did not trigger pod restarts.

> * making calls to K8s API for each and every login request might have security considerations

I am not very familiar with the security implications of calling the K8s API. What would be the possible security considerations?

> (Fabric8 K8s client maintainer here)

Thank you for your work :)

**Comment by andreaTP** [comment]
Hi @snophey and sorry to the delay :sweat: again ...

Let me try to answer.
- The operator will mount the `KafkaUser` Secret in the Registry Pod
- The values of the secret will be propagated by K8s: https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod
- We would need to leverage a mechanism(something like "Dynamic config") in Registry to live update by polling the FileSystem

Alternative:
- the operator can update, using REST calls, Registry after reading the Secret
- the operator can monitor the Secret for updates and reconcile accordingly (an operator can still watch resources even if they are not "owned"

Another one:
- The operator will mount the `KafkaUser` Secret in the Registry Pod
- the operator can monitor the Secret for updates and trigger a reload of credentials, e.g. using an endpoint

> I am not very familiar with the security implications of calling the K8s API. What would be the possible security considerations?

On top of my mind, the number of login requests would linearly impact K8s API Server access, we would need to have in place proper Rate limiting etc. I guess...

An additional question I haven't asked before, this mechanism will be used in production?
Or is it intended to be for dev environments?

**Comment by snophey** [comment]
Thank you for taking the time to explain this, @andreaTP. The missing puzzle piece for me was that Kubernetes automatically updates mounted secrets when changed. In that case, I am open to taking another stab at implementing this (this time, reading the credentials from a file). Do you (and the team) have a preferred approach for this? Polling the filesystem sounds like the simplest solution from an operations standpoint. To expand upon your idea:

- Operator watches for secrets annotated with `strimzi.io/kind: KafkaUser`
- Operator maintains a consolidated "secret file" with all `KafkaUser` credentials and mounts it in each registry (if configured)
- The registry polls the file for changes using something like the `WatchService` API

I'd be happy to submit a new PR implementing this approach if the team agrees that this is the way to go.

> An additional question I haven't asked before, this mechanism will be used in production?

This is intended for both development and production use.

**Comment by andreaTP** [comment]
Thanks for getting back!

> This is intended for both development and production use.

Please, just remember all the shortcomings to use Basic Auth in prod ... 

Your proposed approach looks fine to me, thanks for the iteration.
I would like @EricWittmann feedback on:

> The registry polls the file for changes using something like the WatchService API

Is there a better way of doing it?

**Comment by EricWittmann** [comment]
Hey @snophey - thanks for your patience on this and sorry for taking so long with it.  We've had a few "hair on fire" issues in the last couple of weeks as distractions.  :)

> I'd be happy to submit a new PR implementing this approach if the team agrees that this is the way to go.

After internal discussions, let's go with your proposed approach.  I don't have any strong feelings about an API to use to implement the file polling.  As long as the polling is disabled when this feature is not being used, I'm cool.

**Comment by snophey** [comment]
Hi @EricWittmann, sorry for the delayed response on my part as well. I just returned from holidays. I am closing this PR then. Will re-submit after implementing the proposed approach.

**Pull request closed without merging**